File: core/src/commonMain/kotlin/com/xhsdn/core/url/LinkExtractor.kt
package com.xhsdn.core.url

/**
 * 从用户输入（剪贴板、粘贴、手动输入）中提取 XHS 链接。
 * 复刻 [XHSDownloader.extractLinks] 逻辑（java:592-649），改为纯 Kotlin。
 */
object LinkExtractor {
    private val XHS_LINK_PATTERN = Regex("(?:https?://)?www\\.xiaohongshu\\.com/explore/\\S+")
    private val XHS_USER_PATTERN = Regex("(?:https?://)?www\\.xiaohongshu\\.com/user/profile/[a-z0-9]+/\\S+")
    private val XHS_SHARE_PATTERN = Regex("(?:https?://)?www\\.xiaohongshu\\.com/discovery/item/\\S+")
    private val XHS_SHORT_PATTERN = Regex("(?:https?://)?xhslink\\.com/[^\\s\\\"<>\\\\\\^`{|}，。；！？、【】《》]+")

    fun extractLinks(input: String?): List<String> {
        if (input.isNullOrEmpty()) return emptyList()
        val urls = mutableListOf<String>()
        val parts = input.split("\\s+".toRegex())
        for (part in parts) {
            if (part.isBlank()) continue
            val shortMatcher = XHS_SHORT_PATTERN.find(part)
            if (shortMatcher != null) {
                val shortUrl = part.substring(shortMatcher.range.first, shortMatcher.range.last + 1)
                urls += shortUrl // 短链需要 [ShortLinkResolver] 再 follow redirect 后再决定
                continue
            }
            val shareMatcher = XHS_SHARE_PATTERN.find(part)
            if (shareMatcher != null) {
                urls += part.substring(shareMatcher.range.first, shareMatcher.range.last + 1)
                continue
            }
            val linkMatcher = XHS_LINK_PATTERN.find(part)
            if (linkMatcher != null) {
                urls += part.substring(linkMatcher.range.first, linkMatcher.range.last + 1)
                continue
            }
            val userMatcher = XHS_USER_PATTERN.find(part)
            if (userMatcher != null) {
                urls += part.substring(userMatcher.range.first, userMatcher.range.last + 1)
            }
        }
        return urls
    }
}

File: web/src/core/url/LinkExtractor.ts
/**
 * 从用户输入(剪贴板、粘贴、手动输入)中提取 XHS 链接。
 *
 * 1:1 翻译自 `core/src/commonMain/kotlin/com/xhsdn/core/url/LinkExtractor.kt`。
 * Kotlin `Regex.find(part)` 等价于 TS `regex.exec(part)` —— 返回 null 时即未命中。
 */
const XHS_LINK_PATTERN = /(?:https?:\/\/)?www\.xiaohongshu\.com\/explore\/\S+/;
const XHS_USER_PATTERN = /(?:https?:\/\/)?www\.xiaohongshu\.com\/user\/profile\/[a-z0-9]+\/\S+/;
const XHS_SHARE_PATTERN = /(?:https?:\/\/)?www\.xiaohongshu\.com\/discovery\/item\/\S+/;
const XHS_SHORT_PATTERN = /(?:https?:\/\/)?xhslink\.com\/[^\s"<>\\^`{|}，。；！？、【】《》]+/;

export function extractLinks(input: string | null | undefined): string[] {
  if (!input) return [];
  const urls: string[] = [];
  // Kotlin "\\s+" 等价于 JS /\s+/
  const parts = input.split(/\s+/);
  for (const part of parts) {
    if (!part) continue;

    const shortMatch = part.match(XHS_SHORT_PATTERN);
    if (shortMatch) {
      urls.push(shortMatch[0]);
      continue;
    }
    const shareMatch = part.match(XHS_SHARE_PATTERN);
    if (shareMatch) {
      urls.push(shareMatch[0]);
      continue;
    }
    const linkMatch = part.match(XHS_LINK_PATTERN);
    if (linkMatch) {
      urls.push(linkMatch[0]);
      continue;
    }
    const userMatch = part.match(XHS_USER_PATTERN);
    if (userMatch) {
      urls.push(userMatch[0]);
    }
  }
  return urls;
}
